main.nf

```diff
@@ -3,119 +3,153 @@
-
-        """
-        get_experiment_data.R\
-                --accesssion-code ${params.data_download.ref_acc_code}\
-                --expr-data-type ${params.data_download.expr_data_type}\
-                --normalisation-method ${params.data_download.normalisation_method}\
-                --output-dir-name ${params.data_download.ref_output_dir}\
-                --get-condensed-sdrf\
-                --get-marker-genes\
-                --number-of-clusters ${params.data_download.ref_num_clust}
-
-        # rename files to avoid name collisions in subsequent processes
-        mv ${ref_data}/10x_data ${ref_data}/ref_10x_data
-        mv ${ref_data}/condensed-sdrf.tsv ${ref_data}/ref_sdrf.txt
-        mv ${ref_data}/${ref_markers} ${ref_data}/ref_${ref_markers}
 
-        """
-    }
-    
-    process unmelt_sdrf_ref {
-        conda "${baseDir}/envs/exp_metadata.yaml"
         input:
-            file(condensed_sdrf) from CONDENSED_SDRF_REF
+            tuple file(data), val(dataset_id), val(barcode_col), val(cell_type_col), val(matrix_type) from TRAINING_DATA
 
         output:
-            file("ref_sdrf_proc.tsv") into UNMELT_SDRF_REF
-
+            tuple file("${data}.query"), file("${data}.ref"), val(dataset_id), val(barcode_col), val(cell_type_col), val(matrix_type) into TRAINING_DATA_UNMELT
+	    file("${data}/unmelted_sdrf.tsv") into UNMELTED_SDRF_QUERY
         """
         unmelt_condensed.R\
-                -i ${condensed_sdrf}\
-                -o ref_sdrf_proc.tsv\
-                --retain-types\
-                --has-ontology                 
+                -i ${data}/condensed-sdrf.tsv\
+                -o ${data}/unmelted_sdrf.tsv\
+                --has-ontology\
+                --retain-types ${params.data_import.unmelt_sdrf.retain_types}
+	# rename data dir name to avoid downstream file name collision
+	parallel cp -R ${data} ::: ${data}.query ${data}.ref 
         """
-    }
-    
+    } 
+    TRAINING_DATA_UNMELT.set{ TRAINING_DATA_PROCESSED }
 }else{
-   	REF_10X_DIR = Channel.fromPath(params.ref_10x_dir, checkIfExists: true).first()
-   	QUERY_10X_DIR = Channel.fromPath(params.query_10x_dir, checkIfExists: true).first()
-   	UNMELT_SDRF_REF = Channel.fromPath(params.unmelt_sdrf_ref, checkIfExists: true).first()
-   	UNMELTED_SDRF_QUERY = Channel.fromPath(params.unmelt_sdrf_query, checkIfExists: true).first()
-    	REF_MARKERS = Channel.fromPath(params.ref_markers, checkIfExists: true).first()
+    TRAINING_DATA.set{ TRAINING_DATA_PROCESSED }
+	}
+	
+	// fork queue channel contents into channels for corresponding tools
+	TRAINING_DATA_PROCESSED.into{
+	    GARNETT_DATA
+	    SCMAP_CELL_DATA
+	    SCMAP_CLUSTER_DATA
+	    SCPRED_DATA
+	}
+	
+	// add number of clusters to Garnett data 
+	GARNETT_FULL_DATA = GARNETT_DATA.merge(N_CLUST)
+
+}else{
+	// manual data input (Cross-Validation)
+	
+	MANUAL_INPUT_DATA = Channel.fromPath("${params.input_data}/*.zip")
```

Comment:
so you're assuming that data will be provided as an archive?

Reply:
This is the input coming from the CV workflow, which comes into zipfiles. To make it even more generic I could maybe say: 
```
else{
   if(cross_validation_input == "True"){do what I'm doing}
   if(manual_data_input == "True"){ different structure}
}
```

Reply:
@rubenchazarra If that's not adding too much complexity, go for it. Thanks!